parlai/agents/transformer/modules/encoder.py

```diff
@@ -20,12 +22,75 @@
     MultiHeadAttention,
     TransformerFFN,
 )
+from parlai.agents.transformer.modules.interfaces import ComponentSpec, TComponent
 from parlai.core.opt import Opt
 from parlai.utils.misc import warn_once
 from parlai.utils.torch import PipelineHelper
 
 
-class TransformerEncoder(nn.Module):
+class TransformerEncoderLayer(nn.Module, TComponent):
+    """
+    Implements a single Transformer encoder layer.
+    """
+
+    @dataclass
+    class Manifest(TComponent.Manifest):
+        self_attention: Type[MultiHeadAttention] = MultiHeadAttention
+        feedforward: Type[TransformerFFN] = TransformerFFN
+
+    def __init__(
+        self,
+        n_heads: int,
+        embedding_size: int,
+        ffn_size: int,
+        attention_dropout: float = 0.0,
+        relu_dropout: float = 0.0,
+        dropout: float = 0.0,
+        activation: str = 'relu',
+        variant: Optional[str] = None,
+        manifest: Optional[Manifest] = None,
+    ):
+        super().__init__()
+        manifest = manifest or self.Manifest()
+        self.dim = embedding_size
+        self.ffn_dim = ffn_size
+        self.activation = activation
+        self.variant = variant
+        self.attention = manifest.self_attention(
+            n_heads, embedding_size, dropout=attention_dropout  # --attention-dropout
+        )
+        self.norm1 = torch.nn.LayerNorm(embedding_size, eps=LAYER_NORM_EPS)
+        self.ffn = manifest.feedforward(
+            embedding_size,
+            ffn_size,
+            relu_dropout=relu_dropout,
+            activation=self.activation,
+        )
+        self.norm2 = torch.nn.LayerNorm(embedding_size, eps=LAYER_NORM_EPS)
+        self.dropout = nn.Dropout(p=dropout)
+
+    def forward(self, tensor: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
+        """
+        Forward pass.
+        """
+        residual = tensor
+        if self.variant == 'prelayernorm':
```

Comment:
one could also view the different variants as custom variations, right?

Reply:
For sure! I didn't do anything like that as part of this PR to keep it simple, but I could imagine a pretty simple way to implement that without this `if` statement:
```
class PreLayerNormTransformerEncoderLayer(TransformerEncoderLayer):
  def forward(self, tensor: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
    residual = tensor
    tensor = self.norm1(tensor)
    attended_tensor = self.attention(tensor, mask=mask)[0]
    tensor = residual + self.dropout(attended_tensor)
    residual = tensor
    tensor = self.norm2(tensor)
    tensor = residual + self.dropout(self.ffn(tensor))
    tensor *= mask.unsqueeze(-1).type_as(tensor)
    return tensor

manifest.encoder_layer = ComponentSpec(
  PreLayerNormTransformerEncoderLayer,
  TransformerEncoderLayer.Manifest(),
)
```

To me the code feels simple enough as is for now, but if we end up with a ton of variations then maybe we'll want to split them out into different implementations like this.

Reply:
yes agreed, not in the scope of this current PR but could be an interesting application from it